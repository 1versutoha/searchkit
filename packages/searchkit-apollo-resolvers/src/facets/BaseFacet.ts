import { FilterSet } from '../core/QueryManager'

export interface BaseFacet {
  TYPE: string
  SELECTOR: string
  getId(): string
  getLabel(): string
  getAggregation(): any
  getFilter(filter: FilterSet): any
  transformResponse(response: any): any
}
