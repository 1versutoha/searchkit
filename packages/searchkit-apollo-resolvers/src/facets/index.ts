export { default as MultipleSelectFacet } from './MultipleSelectFacet'
export { default as RefinementSelectFacet } from './RefinementSelectFacet'
