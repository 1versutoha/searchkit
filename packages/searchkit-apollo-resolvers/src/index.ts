export * from './resolvers'
export { default as MultiMatchQuery } from './query/MultiMatchQuery'
