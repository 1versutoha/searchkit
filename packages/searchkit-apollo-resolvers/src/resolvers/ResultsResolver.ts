import QueryManager, { FilterSet } from '../core/QueryManager'
import SearchkitRequest from '../core/SearchkitRequest'

export interface SearchkitConfig {
  host: string
  hits: {
    fields: string[]
  }
}

export interface ResultsResolverParameters {
  filters: Array<FilterSet>
  query: string
}

export default (config: SearchkitConfig) => async (parent, parameters, ctx) => {
  const queryManager = new QueryManager(parameters.filters, parameters.query)
  const skRequest = new SearchkitRequest(queryManager)

  ctx.skRequest = skRequest

  return {}
}
