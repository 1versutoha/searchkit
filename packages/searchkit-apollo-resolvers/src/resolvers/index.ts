export { default as HitsResolver } from './HitsResolver'
export { default as ResultsResolver, SearchkitConfig } from './ResultsResolver'
export { default as FacetsResolver } from './FacetsResolver'
