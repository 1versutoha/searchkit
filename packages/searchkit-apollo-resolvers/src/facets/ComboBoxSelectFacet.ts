import { MixedFilter, ValueFilter } from '../core/QueryManager'
import { BaseFacet, FacetOptions } from './BaseFacet'
import { createRegexQuery } from './utils'

interface ComboBoxSelectFacetConfig {
  id: string
  field: string
  label: string
  display?: 'Combo' | string
  multipleSelect?: boolean
}

class ComboBoxSelectFacet implements BaseFacet {
  public TYPE = 'ComboBoxSelectFacet'
  public excludeOwnFilters = false

  constructor(public config: ComboBoxSelectFacetConfig) {
    this.excludeOwnFilters = this.config.multipleSelect
  }
  getLabel(): string {
    return this.config.label
  }

  getId() {
    return this.config.id
  }

  getFilters(filters: Array<ValueFilter>) {
    const condition = this.excludeOwnFilters ? 'should' : 'must'
    return {
      bool: {
        [condition]: filters.map((term) => ({ term: { [this.config.field]: term.value } }))
      }
    }
  }

  getAggregation(overrides: FacetOptions) {
    if (!overrides) {
      return {}
    }
    return {
      [this.config.id]: {
        terms: {
          field: this.config.field,
          size: overrides?.size || 10,
          ...(overrides?.query ? { include: createRegexQuery(overrides.query) } : {})
        }
      }
    }
  }

  transformResponse(response: any) {
    return {
      id: this.getId(),
      label: this.getLabel(),
      type: this.TYPE,
      display: this.config.display || 'List',
      entries:
        response?.buckets.map((entry) => ({
          id: `${this.getId()}_${entry.key}`,
          label: entry.key,
          count: entry.doc_count
        })) || []
    }
  }
}

export default ComboBoxSelectFacet
