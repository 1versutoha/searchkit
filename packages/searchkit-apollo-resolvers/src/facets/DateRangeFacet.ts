import { RangeFilter, DateRangeFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface DateRangeFacetConfig {
  id: string
  field: string
  label: string
  size?: number
  display?: 'DateRange' | string
  range: {
    min: number
    max: number
    interval: number
    min_doc_count?: number
  }
}

class DateRangeFacet implements BaseFacet {
  public TYPE = 'DateRangeFacet'
  public excludeOwnFilters = true

  constructor(public config: DateRangeFacetConfig) {}
  getLabel(): string {
    return this.config.label
  }

  getId() {
    return this.config.id
  }

  getFilter(filter: DateRangeFilter) {
    return { range: { [this.config.field]: { gte: filter.dateMin, lte: filter.dateMax } } }
  }

  getAggregation() {
    return {}
  }

  transformResponse() {
    return {
      id: this.getId(),
      label: this.getLabel(),
      type: this.TYPE,
      display: this.config.display || 'DateRange',
      entries: []
    }
  }
}

export default DateRangeFacet
