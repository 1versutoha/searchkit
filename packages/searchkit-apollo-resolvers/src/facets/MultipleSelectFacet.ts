import { BaseFacet } from './BaseFacet'
import { FilterSet } from '../core/QueryManager'

interface MultipleSelectFacetConfig {
  id: string
  field: string
  size?: number
  label: string
}

class MultipleSelectFacet implements BaseFacet {
  public TYPE = 'MultipleSelectFacet'
  public SELECTOR = 'OR'

  constructor(public config: MultipleSelectFacetConfig) {}
  getLabel(): string {
    return this.config.label
  }

  getId() {
    return this.config.id
  }

  getFilter(filter: FilterSet) {
    return {
      bool: {
        must: filter.selected.map((term) => ({ term: { [this.config.field]: term } }))
      }
    }
  }

  getAggregation() {
    return {
      [this.config.id]: {
        terms: {
          field: this.config.field,
          size: this.config.size || 5
        }
      }
    }
  }

  transformResponse(response: any) {
    return {
      id: this.getId(),
      label: this.getLabel(),
      type: this.TYPE,
      entries: response.buckets.map((entry) => ({
        id: entry.key,
        label: entry.key,
        count: entry.doc_count
      }))
    }
  }
}

export default MultipleSelectFacet
