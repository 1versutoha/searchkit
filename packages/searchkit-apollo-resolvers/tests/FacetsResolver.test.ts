import { SearchkitConfig } from '../src/resolvers/ResultsResolver'
import { MultiMatchQuery } from '../src'
import { MultipleSelectFacet, RefinementSelectFacet } from '../src/facets'
import { setupTestServer, callQuery } from './support/helper'

describe('Hits Resolver', () => {
  describe('should return as expected', () => {
    const runQuery = async (gql) => {
      const response = await callQuery({ gql })
      return response
    }

    const config: SearchkitConfig = {
      host: 'http://localhost:9200/movies/_search',
      hits: {
        fields: ['actors', 'writers']
      },
      query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] }),
      facets: [
        new MultipleSelectFacet({ id: 'writers', field: 'writers.raw', label: 'Writers' }),
        new RefinementSelectFacet({ id: 'actors', field: 'actors.raw', label: 'Actors' }),
        new RefinementSelectFacet({ id: 'type', field: 'type.raw', label: 'Type' }),
        new RefinementSelectFacet({ id: 'genres', field: 'genres.raw', label: 'Genres' })
      ]
    }

    it('should return correct Results', async () => {
      setupTestServer(config)

      const gql = `
        {
          results(query: "") {
            hits(page: {rows: 10, start: 0 }) {
              id
            }
            facets {
              id
              type
              label
              entries {
                id
                count
                label
              }
            }
          }
        }
      `

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })

    it('should return correct results with one filter', async () => {
      setupTestServer(config)

      const gql = `
        {
          results(
            query: "",
            filters: [
              { id: "writers", selected: ["Damon Lindelof"]}
            ]
            ) {
            hits(page: {rows: 10, start: 0 }) {
              id
            }
            facets {
              id
              type
              label
              entries {
                id
                count
                label
              }
            }
          }
        }
      `

      const response = await runQuery(gql)

      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })
  })
})
