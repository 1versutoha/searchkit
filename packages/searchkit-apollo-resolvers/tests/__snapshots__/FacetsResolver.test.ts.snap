// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hits Resolver should return as expected should return correct Results 1`] = `
Object {
  "results": Object {
    "facets": Array [
      Object {
        "entries": Array [
          Object {
            "count": 83,
            "id": "J.J. Abrams",
            "label": "J.J. Abrams",
          },
          Object {
            "count": 74,
            "id": "Jeffrey Lieber",
            "label": "Jeffrey Lieber",
          },
          Object {
            "count": 73,
            "id": "Damon Lindelof",
            "label": "Damon Lindelof",
          },
          Object {
            "count": 53,
            "id": "James Manos Jr.",
            "label": "James Manos Jr.",
          },
          Object {
            "count": 53,
            "id": "Jeff Lindsay",
            "label": "Jeff Lindsay",
          },
        ],
        "id": "writers",
        "label": "Writers",
        "type": "MultipleSelectFacet",
      },
      Object {
        "entries": Array [
          Object {
            "count": 73,
            "id": "Naveen Andrews",
            "label": "Naveen Andrews",
          },
          Object {
            "count": 56,
            "id": "Jennifer Carpenter",
            "label": "Jennifer Carpenter",
          },
          Object {
            "count": 56,
            "id": "Michael C. Hall",
            "label": "Michael C. Hall",
          },
          Object {
            "count": 53,
            "id": "Emilie de Ravin",
            "label": "Emilie de Ravin",
          },
          Object {
            "count": 49,
            "id": "Jared Padalecki",
            "label": "Jared Padalecki",
          },
        ],
        "id": "actors",
        "label": "Actors",
        "type": "RefinementSelectFacet",
      },
      Object {
        "entries": Array [
          Object {
            "count": 3120,
            "id": "Movie",
            "label": "Movie",
          },
          Object {
            "count": 521,
            "id": "Episode",
            "label": "Episode",
          },
          Object {
            "count": 473,
            "id": "Series",
            "label": "Series",
          },
          Object {
            "count": 48,
            "id": "Game",
            "label": "Game",
          },
        ],
        "id": "type",
        "label": "Type",
        "type": "RefinementSelectFacet",
      },
      Object {
        "entries": Array [
          Object {
            "count": 2529,
            "id": "Drama",
            "label": "Drama",
          },
          Object {
            "count": 1274,
            "id": "Comedy",
            "label": "Comedy",
          },
          Object {
            "count": 786,
            "id": "Crime",
            "label": "Crime",
          },
          Object {
            "count": 725,
            "id": "Action",
            "label": "Action",
          },
          Object {
            "count": 666,
            "id": "Adventure",
            "label": "Adventure",
          },
        ],
        "id": "genres",
        "label": "Genres",
        "type": "RefinementSelectFacet",
      },
    ],
    "hits": Array [
      Object {
        "id": "tt0111161",
      },
      Object {
        "id": "tt0068646",
      },
      Object {
        "id": "tt0468569",
      },
      Object {
        "id": "tt0944947",
      },
      Object {
        "id": "tt0110912",
      },
      Object {
        "id": "tt0137523",
      },
      Object {
        "id": "tt0167260",
      },
      Object {
        "id": "tt1375666",
      },
      Object {
        "id": "tt0120737",
      },
      Object {
        "id": "tt0133093",
      },
    ],
  },
}
`;

exports[`Hits Resolver should return as expected should return correct results with one filter 1`] = `
Object {
  "results": Object {
    "facets": Array [
      Object {
        "entries": Array [
          Object {
            "count": 83,
            "id": "J.J. Abrams",
            "label": "J.J. Abrams",
          },
          Object {
            "count": 74,
            "id": "Jeffrey Lieber",
            "label": "Jeffrey Lieber",
          },
          Object {
            "count": 73,
            "id": "Damon Lindelof",
            "label": "Damon Lindelof",
          },
          Object {
            "count": 53,
            "id": "James Manos Jr.",
            "label": "James Manos Jr.",
          },
          Object {
            "count": 53,
            "id": "Jeff Lindsay",
            "label": "Jeff Lindsay",
          },
        ],
        "id": "writers",
        "label": "Writers",
        "type": "MultipleSelectFacet",
      },
      Object {
        "entries": Array [
          Object {
            "count": 70,
            "id": "Naveen Andrews",
            "label": "Naveen Andrews",
          },
          Object {
            "count": 51,
            "id": "Emilie de Ravin",
            "label": "Emilie de Ravin",
          },
          Object {
            "count": 36,
            "id": "Henry Ian Cusick",
            "label": "Henry Ian Cusick",
          },
          Object {
            "count": 34,
            "id": "Matthew Fox",
            "label": "Matthew Fox",
          },
          Object {
            "count": 20,
            "id": "Jorge Garcia",
            "label": "Jorge Garcia",
          },
        ],
        "id": "actors",
        "label": "Actors",
        "type": "RefinementSelectFacet",
      },
      Object {
        "entries": Array [
          Object {
            "count": 69,
            "id": "Episode",
            "label": "Episode",
          },
          Object {
            "count": 3,
            "id": "Movie",
            "label": "Movie",
          },
          Object {
            "count": 1,
            "id": "Series",
            "label": "Series",
          },
        ],
        "id": "type",
        "label": "Type",
        "type": "RefinementSelectFacet",
      },
      Object {
        "entries": Array [
          Object {
            "count": 73,
            "id": "Adventure",
            "label": "Adventure",
          },
          Object {
            "count": 70,
            "id": "Drama",
            "label": "Drama",
          },
          Object {
            "count": 70,
            "id": "Fantasy",
            "label": "Fantasy",
          },
          Object {
            "count": 2,
            "id": "Action",
            "label": "Action",
          },
          Object {
            "count": 2,
            "id": "Sci-Fi",
            "label": "Sci-Fi",
          },
        ],
        "id": "genres",
        "label": "Genres",
        "type": "RefinementSelectFacet",
      },
    ],
    "hits": Array [
      Object {
        "id": "tt0411008",
      },
      Object {
        "id": "tt1408101",
      },
      Object {
        "id": "tt1446714",
      },
      Object {
        "id": "tt0816711",
      },
      Object {
        "id": "tt0636272",
      },
      Object {
        "id": "tt0994359",
      },
      Object {
        "id": "tt1453335",
      },
      Object {
        "id": "tt0636275",
      },
      Object {
        "id": "tt1321692",
      },
      Object {
        "id": "tt0636277",
      },
    ],
  },
}
`;
